spec/wrap.md

```diff
@@ -0,0 +1,104 @@
+# WRAP Specification
+
+WRAP defines the rules for bi-directional communication between a Host and a Wasm module by defining the processes of invocation and subinvocation. It makes use of Host allocated Wasm shared memory to pass encoded buffers between the Host and the Wasm module.
+
+The Wasm module must define the following imports and exports:
+
+#### Exports
+- `_invoke: (bufferLen: u32) => u32`
+    - sends the length of input buffer to Wasm module and it returns the length and the ptr to the new 8 byte long response buffer which contains length of the result buffer + pointer to the result buffer
+
+#### Imports
+- `__subinvoke: (bufferPtr: u32, bufferLen: u32) => u32`
+    - Allows sending any buffer from the Wasm module to the host. Host returns the length of the result buffer to the Wasm module.
+- `__fill_buffer: (bufferPtr: u32) => void` 
```

Comment:
This function feels too generic, and lacks context as to "what buffer is being filled". I think we could add a "bufferId" alongside the "bufferPtr" here. Also would need to add "bufferId" alongside the "bufferLen" in the _invoke export.

Reply:
Reworded this a bit. But it still a generic functions, it can be any buffer that the host is sending to the wasm module. It doesn't need an ID to identify it because there can be only one type depending on context

Reply:
This function can also be named `load_buffer` 
What do you guys prefer: `fill_buffer 🍏` vs `load_buffer 🍎`